### PR TITLE
PERF-5212 Add new $in workload that returns documents

### PR DESCRIPTION
**Pull request by blagoev**
**Jira Ticket:** [PERF-5212](https://jira.mongodb.org/browse/PERF-5212)

Description: A new $in workload that matches and returns documents.

**Patch testing results:**  

If applicable, link a patch test showing code changes running successfully

**Related Issues:**  
https://jira.mongodb.org/browse/DEVPROD-5806

### CODE REVIEW

File: src/value_generators/src/DocumentGenerator.cpp [hunk 1]
```diff
@@ -1949,12 +1949,13 @@ class ObjectGenerator : public Generator<bsoncxx::document::value> {
...
 };
 
-
 class IncGenerator : public Generator<int64_t> {
 public:
     IncGenerator(const Node& node, GeneratorArgs generatorArgs)
         : _step{node["step"].maybe<int64_t>().value_or(1)} {
-        _counter = node["start"].maybe<int64_t>().value_or(1) +
+       
+        int64_t start = node["start"] ? intGenerator(node["start"], generatorArgs)->evaluate() : 1;
```
Comment (jimoleary) [inline src/value_generators/src/DocumentGenerator.cpp:1957]: Can you add some tests and examples?

Generally we add tests [here](https://github.com/mongodb/genny/blob/master/src/value_generators/test/DocumentGeneratorTestCases.yml#L3) and add document some examples [here](https://github.com/mongodb/genny/blob/master/src/workloads/docs/Generators.yml#L57).

File: src/workloads/scale/LargeIndexedInsMatchingDocuments.yml [hunk 2]
```diff
@@ -0,0 +1,84 @@
+SchemaVersion: 2018-07-01
+Owner: "@mongodb/product-perf"
+Description: |
+  This is an indexed $in workload that matches documents in a collection.
+  This workload is intended to test a more representative workload
+
+  An extremelly common usecase for users is to fetch objects and hydrate with follow up $in query
```
Comment (jimoleary) [inline src/workloads/scale/LargeIndexedInsMatchingDocuments.yml:7]: nit: typo ```extremelly```.

File: src/workloads/scale/LargeIndexedInsMatchingDocuments.yml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    name: "Tyler",
+    posts: [ObjectId(1), ObjectId(2), ObjectId(3)],
+  }
+
+  The second query fetches the related, but not embedded/denormalized objects
+
+  { $in: [ObjectId(1), ObjectId(2), ObjectId(3)] }
+
+  This workload focuses on the second query of this use case matching documents
+  using arrays of integers that are generated to be used for the $in queries.
```
Comment (jimoleary) [inline src/workloads/scale/LargeIndexedInsMatchingDocuments.yml:22]: Can you explain what metrics are important for this workload?

File: src/workloads/scale/LargeIndexedInsMatchingDocuments.yml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+Clients:
+  Default:
+    QueryOptions:
+      maxPoolSize: 20000
+  Update:
```
Comment (jimoleary) [inline src/workloads/scale/LargeIndexedInsMatchingDocuments.yml:27]: Do you really need the **Clients** at all here?

You're not using the Update pool and you are no where near 20_000 clients elsewhere.

Note: I assume this is just cut'n'pasta from LargeIndexedIns.yml but I don't think it is required.

File: src/workloads/scale/LargeIndexedInsMatchingDocuments.yml [hunk 5]
```diff
@@ -0,0 +1,84 @@
+SchemaVersion: 2018-07-01
+Owner: "@mongodb/product-perf"
+Description: |
```
Comment (jimoleary) [inline src/workloads/scale/LargeIndexedInsMatchingDocuments.yml:3]: Can you add a **keywords:** section?

File: src/workloads/scale/LargeIndexedInsMatchingDocuments.yml [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+              Filter:
+                _id:
+                  {
+                    ^Array:
+                      {
+                        of:
+                          {
+                            ^Inc:
+                              {
+                                start: { ^RandomInt: { min: 0, max: 2028000 } }
```
Comment (jimoleary) [inline src/workloads/scale/LargeIndexedInsMatchingDocuments.yml:69]: Is 2028000 significant? Can you document the rationale somewhere if it is.

File: src/workloads/scale/LargeIndexedInsMatchingDocuments.yml [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+                              },
+                          },
+                        number: 20000,
```
Comment (jimoleary) [inline src/workloads/scale/LargeIndexedInsMatchingDocuments.yml:72]: Wow, this is likely to be a very heavy weight test!

Have you tested it to see how noisy it is likely to be or if it generates stable results?

File: src/workloads/scale/LargeIndexedInsMatchingDocuments.yml [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    Type: Loader
+    Threads: 1
+    Phases:
+      - Repeat: 1
+        Database: indexedins
+        CollectionCount: 1
+        Threads: 1
+        DocumentCount: 2048000
+        BatchSize: 100
+        Document:
```
Comment (jimoleary) [inline src/workloads/scale/LargeIndexedInsMatchingDocuments.yml:42]: The documents work out ~57 bytes:

```
Object.bsonsize({_id: 2048000, title: '12345678', text: '12345678'})
```

2,048,000 works out at about 0.1 GB, is this a big enough test?

**Review by jimoleary — COMMENTED** [review]
Mostly minor nits / requests.

File: src/workloads/scale/LargeIndexedInsMatchingDocuments.yml (hunk 4 above)
Comment (thessem) [reply src/workloads/scale/LargeIndexedInsMatchingDocuments.yml:27]: By default we will be grabbing up to `maxPoolSize` clients on boot, so 20K might need to be justified

Comment (blagoev) [reply src/workloads/scale/LargeIndexedInsMatchingDocuments.yml:27]: @jimoleary is right. This is a copy/paste from `LargeIndexedIns` workload. I will remove it.

File: src/value_generators/src/DocumentGenerator.cpp (hunk 1 above)
Comment (blagoev) [reply src/value_generators/src/DocumentGenerator.cpp:1957]: thanks for the reminder, will add tests and examples

Comment (blagoev) [reply src/value_generators/src/DocumentGenerator.cpp:1957]: added

File: src/workloads/scale/LargeIndexedInsMatchingDocuments.yml (hunk 6 above)
Comment (blagoev) [reply src/workloads/scale/LargeIndexedInsMatchingDocuments.yml]: added comments

File: src/workloads/scale/LargeIndexedInsMatchingDocuments.yml (hunk 8 above)
Comment (blagoev) [reply src/workloads/scale/LargeIndexedInsMatchingDocuments.yml:51]: we discussed internally and settled on around 2,000,000 value for the document count. We think it should cover the workload needs. 
We can increase if we notice it's not enough.

File: src/workloads/scale/LargeIndexedInsMatchingDocuments.yml [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    Threads: 1
+    Phases:
+      - Repeat: 1
+        Database: *database
+        CollectionCount: 1
+        Threads: 1
+        DocumentCount: *documentCount
+        BatchSize: 100
+        Document:
+          _id: { ^Inc: { start: 0 } }
```
Comment (brad-devlugt) [inline src/workloads/scale/LargeIndexedInsMatchingDocuments.yml:52]: Are the performance characteristics different if we use ObjectIDs?

**Review by jimoleary — COMMENTED** [review]
a few nits and a question about idhack

File: src/workloads/scale/LargeIndexedInsMatchingDocuments.yml [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+                          # An array of distinct and sequential integers up to documentCount. It will make sure all documents in this filter are found.
+                          ^Array: 
+                            {
+                              of:
+                                {
+                                  ^Inc:
+                                    {
+                                      start:
+                                        {
+                                          ^RandomInt: { min: 0, max: 2028000 }, # max ^Inc.start is documentCount - inArraySize
```
[the remark this answers is not in the document]
Comment (blagoev) [reply src/workloads/scale/LargeIndexedInsMatchingDocuments.yml:84]: good catch. I renamed the variable and forgot to change it here.

File: docs/developing.md [hunk 11]
```diff
@@ -54,6 +54,45 @@ Read more about what parameters you can pass [here][catch2].
 
 [catch2]: https://github.com/catchorg/Catch2/blob/v2.5.0/docs/command-line.md#specifying-which-tests-to-run
 
+### Value Generator Tests
+
+When modifing or creating new value generators it is highly valuable and important to test the implementation with a test case and to provide an example of the generator usage.
+
+#### Authroing Test Cases
```
Comment (dalyd) [inline docs/developing.md:61]: s/Authroing/Authoring/

File: docs/developing.md [hunk 12]
```diff
@@ -54,6 +54,45 @@ Read more about what parameters you can pass [here][catch2].
...
+      <template>
+    ThenReturns:
+    - <expected output 1>
+    - <expected output 2>
+    - <expected output 3>
+```
+
+   Note:
+
+   *  The tests are executed in iterations, one for each of the expected outpus.
```
Comment (dalyd) [inline docs/developing.md:77]: s/outpus/outputs/

File: docs/developing.md [hunk 13]
```diff
@@ -54,6 +54,45 @@ Read more about what parameters you can pass [here][catch2].
...
+    ThenReturns:
+    - <expected output 1>
+    - <expected output 2>
+    - <expected output 3>
+```
+
+   Note:
+
+   *  The tests are executed in iterations, one for each of the expected outpus. 
+   For each iteration the `GivenTemplate` is compiled and evaulated with the declared
```
Comment (dalyd) [inline docs/developing.md:78]: s/evaulated/evaluated/

File: docs/developing.md [hunk 14]
```diff
@@ -54,6 +54,45 @@ Read more about what parameters you can pass [here][catch2].
...
+   value generators in it. Then the actual and expected outputs are compared.
+   * In the sample above the template will be evaluated 3 times and the test will run in 3 iterations.
+   * Random generators declared in the templates are created only once and reused for all iterations. The produced values are incremental.
```
Comment (dalyd) [inline docs/developing.md:81]: Do you want to say something about the data being random by reproducible. So when adding a test you may need to run the test first to get results, verify they are reasonable results, and then put them into the expected value?

File: docs/developing.md [hunk 15]
```diff
@@ -54,6 +54,45 @@ Read more about what parameters you can pass [here][catch2].
...
+
+```sh
+~/workspace/genny$ cd build/src/value_generators/
+~/workspace/genny/build/src/value_generators$ ./value_generators_test -d yes --order lex
+```
+
+#### Providing Value Generator Examples
+
+While tests are usually the best documentation, sometimes they may be too verbose and go into edge cases, so clear and concise examples are prefered.
+Value generator examples are documented in src/workloads/docs/Generators.yml.
```
Comment (dalyd) [inline docs/developing.md:95]: Please add a examples to that file when adding or extending a generator.

**Review by dalyd — CHANGES_REQUESTED** [review]

**Review by jimoleary — COMMENTED** [review]
I'm mostly good on the content here ... I have a few nits, questions and suggestions but they are all pretty minor at this stage.

Great work and thanks for adding the documentation !

However I have a question relating to the t2 output, there is a ramp up in the queries? 

Any idea what is the cause of this, do we need a warmup phase or is this a client side limitation?

See the following image 

ramp

File: src/value_generators/test/DocumentGeneratorTestCases.yml [hunk 16]
```diff
@@ -1715,4 +1715,16 @@ Tests:
   - Name: Invalid conversion from string
     GivenTemplate:
       stoi: {^ConvertToInt32: {from: "abc"}}
-    ThenThrows: InvalidValueGeneratorSyntax
\ No newline at end of file
+    ThenThrows: InvalidValueGeneratorSyntax
+
+  - Name: IncrementWithRandomStart
+    GivenTemplate:
+      arrayOfInts: { ^Array: {of: { ^Inc: { start: { ^RandomInt: {min: 10000000000, max: 10000000010}}}}, number: 5}}
```
[the remark this answers is not in the document]
Comment (blagoev) [reply src/value_generators/test/DocumentGeneratorTestCases.yml:1722]: this won't work. it throws 
```
 std::exception::what: Wanted map got 3: [10000000001, 10000000002,
  10000000003, 10000000004, 10000000005]
```

File: src/workloads/scale/LargeIndexedInsMatchingDocuments.yml [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+      - { Nop: true }
+      - { Nop: true }
+
+  - Name: FindBlogPostsById
+    Type: CrudActor
+    Threads: 64
+    Database: *database
+    Phases:
+      - { Nop: true }
+      - Name: MatchDocumentsUsingInById
```
[the remark this answers is not in the document]
Comment (blagoev) [reply src/workloads/scale/LargeIndexedInsMatchingDocuments.yml:68]: well `GetXYZById` means the `_id` of the document. meaning by `the primary key`. and then `GetXyZByName` is just by field `Name`. At least this is what I am used to when doing CRUD implementations.

File: docs/developing.md [hunk 18]
```diff
@@ -54,6 +54,48 @@ Read more about what parameters you can pass [here][catch2].
...
+     When adding a test with random data you may need to run the tests first, validate the results are reasonable and use them as expected values.
+   * Use large numbers for number parameters because short numbers get automatically narrowed to `int32` values whereas number value-generators always produce `int64` values.
+
+*  Testing value generators locally
+
+If you have successfully installed genny with `./run-genny install` you should have a test executable for running locally the value generator tests. Test can be executed with 
+
+```sh
+~/workspace/genny$ cd build/src/value_generators/
+~/workspace/genny/build/src/value_generators$ ./value_generators_test -d yes --order lex
```
[the remark this answers is not in the document]
Comment (blagoev) [reply docs/developing.md:91]: I will add it,

File: docs/developing.md [hunk 19]
```diff
@@ -54,6 +54,48 @@ Read more about what parameters you can pass [here][catch2].
...
+   Note:
+
+   *  The tests are executed in iterations, one for each of the expected outputs. 
+   For each iteration the `GivenTemplate` is compiled and evaluated with the declared
+   value generators in it. Then the actual and expected outputs are compared.
+   * In the sample above the template will be evaluated 3 times and the test will run in 3 iterations.
+   * Random generators declared in the templates are created only once and reused for all iterations. The produced values are incremental. 
+     When adding a test with random data you may need to run the tests first, validate the results are reasonable and use them as expected values.
+   * Use large numbers for number parameters because short numbers get automatically narrowed to `int32` values whereas number value-generators always produce `int64` values.
+
```
[the remark this answers is not in the document]
Comment (blagoev) [reply docs/developing.md:84]: Are we sure about this. I tried moving the new tests on a random place and all tests succeeded. Isn't the random number generator created every time before a test case is run?
here https://github.com/mongodb/genny/blob/master/src/value_generators/test/DocGenYamlTestCaseRunner.cpp#L94

**Comment by blagoev** [comment]
@jimoleary I think the ramp up is caused by the client generating all the find $in random arrays

File: docs/developing.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (jimoleary) [reply docs/developing.md:84]: It looks like you are right ... cool.

File: src/value_generators/test/DocumentGeneratorTestCases.yml (hunk 16 above)
[the remark this answers is not in the document]
Comment (jimoleary) [reply src/value_generators/test/DocumentGeneratorTestCases.yml:1722]: kk.

File: src/workloads/scale/LargeIndexedInsMatchingDocuments.yml (hunk 17 above)
[the remark this answers is not in the document]
Comment (jimoleary) [reply src/workloads/scale/LargeIndexedInsMatchingDocuments.yml:68]: are ```MatchDocumentsUsingInByObjectId``` and ```MatchDocumentsUsingInById``` inconsistent then? The Object Id one is also on ```_id```.

**Comment by jimoleary** [comment]
> @jimoleary I think the ramp up is caused by the client generating all the find $in random arrays


I think that is likely too but does that mean that the test is client limited?

File: src/workloads/scale/LargeIndexedInsMatchingDocuments.yml (hunk 17 above)
[the remark this answers is not in the document]
Comment (jimoleary) [reply src/workloads/scale/LargeIndexedInsMatchingDocuments.yml:68]: doh ... trying to do too many things at once ... my comment  is wrong. resolving.

**Comment by blagoev** [comment]
@dalyd I think I have handled all the comments you had. Could you re-review.

**Review by dalyd — APPROVED** [review]
LGTM to changes in docs/developing.md. Thank you for adding that!

**Review by jimoleary — APPROVED** [review]
LGTM on the change but do we have a new patch with FixedGeneratedValue, has this  removed the ramp up in operations?

**Comment by blagoev** [comment]
@jimoleary Yes the ramp up is gone with pre-generated arrays. Just shared a doc with you with the results.

**Review by jimoleary — APPROVED** [review]
LGTM 
Thanks for sharing the doc on the new ramp up.

**Comment by blagoev** [comment]
Thanks Jim, I will be merging this.
